### PR TITLE
    ipatests: test_epn: enhance CLI testing

### PR DESCRIPTION
**Pull request by fcami**
Enhance test_EPN_nbdays so that it checks:
* that no emails get sent when using --dry-run
* that --from-nbdays implies --dry-run
* that --to-nbdays requires --from-nbdays
    
Signed-off-by: François Cami <fcami@redhat.com>

### CODE REVIEW

**Comment by rcritten** [comment]
Would it make sense to test illegal values for the number of days?

**Comment by fcami** [comment]
It would, yes, thanks for the suggestion.

**Comment by fcami** [comment]
good run: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/037f89be-d730-11ea-a8a5-fa163e168469/
removing WIP

File: ipatests/test_integration/test_epn.py [hunk 1]
```diff
@@ -344,22 +349,114 @@ def test_EPN_smoketest_3(self):
...
+            (stdout_text_client, unused, unused) = self._check_epn_output(
+                self.clients[0], to_nbdays=i
+            )
+            user_list = [user["uid"] for user in json.loads(stdout_text_client)]
+            assert len(user_list) == 1
+            assert user_list[0] == "user{id}".format(id=i - 1)
+
+            # make sure no emails were sent
+            result = self.clients[0].run_command(['ls', '-lha', '/var/mail/'])
+            assert ".." in result.stdout_text
```
Comment (rcritten) [inline ipatests/test_integration/test_epn.py:383]: This just asserts that there is a parent directory, what does that prove?

File: ipatests/test_integration/test_epn.py [hunk 2]
```diff
@@ -344,22 +349,114 @@ def test_EPN_smoketest_3(self):
...
+
+    # Test the to/from nbdays options behavior with illegal input
+
+    def test_EPN_nbdays_input_0(self):
+        """Make sure that --to-nbdays implies --dry-run."""
+
+        (stdout_text_client, unused, unused) = self._check_epn_output(
+            self.clients[0], to_nbdays=5, dry_run=False
+        )
+        assert len(json.loads(stdout_text_client)) >= 1
```
Comment (rcritten) [inline ipatests/test_integration/test_epn.py:412]: Do we care how many lines were returned, what is in those lines or whether the output is valid json?

Comment (fcami) [reply ipatests/test_integration/test_epn.py:412]: That the output is valid json and contains at least one entry: essentially "Make sure that --to-nbdays implies --dry-run."

File: ipatests/test_integration/test_epn.py (hunk 1 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py:383]: Nothing, will fix, thanks for the catch!

File: ipatests/test_integration/test_epn.py (hunk 2 above)
Comment (rcritten) [reply ipatests/test_integration/test_epn.py:412]: I think the test would be better if it verified that the expected output matches what is seen.

Comment (fcami) [reply ipatests/test_integration/test_epn.py:412]: That's more or less what test_EPN_nbdays_1 does. Do you want it in test_EPN_nbdays_input_0 too?

Comment (rcritten) [reply ipatests/test_integration/test_epn.py:412]: I think think that depending on the length of the output and not the content of the output is fragile.

Comment (fcami) [reply ipatests/test_integration/test_epn.py:412]: OK, agreed, will add the required logic there too.

File: ipatests/test_integration/test_epn.py (hunk 1 above)
Comment (rcritten) [reply ipatests/test_integration/test_epn.py:383]: The ls will raise an exception if execution fails so I think you're safe without the assert.

Comment (fcami) [reply ipatests/test_integration/test_epn.py]: OK, I've removed it.

File: ipatests/test_integration/test_epn.py (hunk 2 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py]: Done.

File: ipatests/test_integration/test_epn.py [hunk 3]
```diff
@@ -15,6 +15,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+# W0128 trips on "unused" even if "dummy-variables-rgx=(_.+|unused)"
+# pylint: disable=W0128
+
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:20]: pylint complains when there is a double-assignment to unused in the same line. You can remove this disable line and replace `(stdout_text, unused, unused)` with `(stdout_text, unused, _unused)`.

File: ipatests/test_integration/test_epn.py [hunk 4]
```diff
@@ -244,12 +249,12 @@ def test_EPN_smoketest_1(self):
         ''')
         self.master.put_file_contents('/etc/ipa/epn.conf', epn_conf)
         # check EPN on client (LDAP+GSSAPI)
-        (stdout_text, unused) = self._check_epn_output(
+        (stdout_text, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:252]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py [hunk 5]
```diff
@@ -244,12 +249,12 @@ def test_EPN_smoketest_1(self):
...
             self.clients[0], dry_run=True
         )
         assert len(json.loads(stdout_text)) == 0
         # check EPN on master (LDAPI)
-        (stdout_text, unused) = self._check_epn_output(
+        (stdout_text, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:257]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py [hunk 6]
```diff
@@ -292,10 +297,10 @@ def test_EPN_smoketest_2(self, cleanupusers):
                 ),
             ],
         )
-        (stdout_text_client, unused) = self._check_epn_output(
+        (stdout_text_client, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:300]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py [hunk 7]
```diff
@@ -292,10 +297,10 @@ def test_EPN_smoketest_2(self, cleanupusers):
...
             self.clients[0], dry_run=True
         )
-        (stdout_text_master, unused) = self._check_epn_output(
+        (stdout_text_master, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:303]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py [hunk 8]
```diff
@@ -331,10 +336,10 @@ def test_EPN_smoketest_3(self):
                 password=None,
             )
 
-        (stdout_text_client, unused) = self._check_epn_output(
+        (stdout_text_client, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:339]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py [hunk 9]
```diff
@@ -331,10 +336,10 @@ def test_EPN_smoketest_3(self):
...
             self.clients[0], dry_run=True
         )
-        (stdout_text_master, unused) = self._check_epn_output(
+        (stdout_text_master, unused, unused) = self._check_epn_output(
```
Comment (flo-renaud) [inline ipatests/test_integration/test_epn.py:342]: replace the second `unused` with a different variable name like `_unused`.

File: ipatests/test_integration/test_epn.py (hunk 3 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py:24]: Thanks for the hint. Somehow I find that really messy here, but I've changed to that.

File: ipatests/test_integration/test_epn.py [hunk 10]
```diff
@@ -344,22 +353,117 @@ def test_EPN_smoketest_3(self):
         expected_users = ["user1", "user3", "user7", "user14", "user28"]
         assert sorted(user_lst) == sorted(expected_users)
 
-    def test_EPN_nbdays(self):
+    def test_EPN_nbdays_0(self, cleanupmail):
```
Comment (rcritten) [inline ipatests/test_integration/test_epn.py:356]: Is cleanupmail necessary here or is it to prevent cascading failures?

File: ipatests/test_integration/test_epn.py [hunk 11]
```diff
@@ -344,22 +353,117 @@ def test_EPN_smoketest_3(self):
...
+            )
+            user_list = [user["uid"] for user in json.loads(stdout_text_client)]
+            for user in expected_user_list:
+                assert user in user_list
+            for user in user_list:
+                assert user in expected_user_list
+
+    # Test the to/from nbdays options behavior with illegal input
+
+    def test_EPN_nbdays_input_0(self):
```
Comment (rcritten) [inline ipatests/test_integration/test_epn.py:409]: If you use cleanupmail for the other --dry-run tests shouldn't it be here for consistency? Or in none of them?

File: ipatests/test_integration/test_epn.py [hunk 12]
```diff
@@ -344,22 +353,117 @@ def test_EPN_smoketest_3(self):
...
+           expected user.
+        """
+
+        (stdout_text_client, unused, _unused) = self._check_epn_output(
+            self.clients[0], to_nbdays=5, dry_run=False
+        )
+        assert len(json.loads(stdout_text_client)) == 1
+        assert json.loads(stdout_text_client)[0]["uid"] == "user4"
+
+    def test_EPN_nbdays_input_1(self, cleanupmail):
```
Comment (rcritten) [inline ipatests/test_integration/test_epn.py:421]: This is expected to fail, no cleanupmail necessary

File: ipatests/test_integration/test_epn.py (hunk 10 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py:356]: It is necessary because if we add another test before it that generates emails we need to make sure /var/mail is empty before this starts.

File: ipatests/test_integration/test_epn.py (hunk 11 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py:409]: I don't think so, because test_EPN_nbdays_0 would trip if --dry-run didn't work as expected.

File: ipatests/test_integration/test_epn.py (hunk 12 above)
Comment (fcami) [reply ipatests/test_integration/test_epn.py:421]: agreed & changed.

**Comment by flo-renaud** [comment]
@fcami thanks for the update, LGTM. You can remove the temp commit and I will ack when PRCI completes.

**Comment by fcami** [comment]
Thanks @flo-renaud & done

**Comment by flo-renaud** [comment]
Tests are green, ACK.

**Comment by flo-renaud** [comment]
master:

* e1750e2a18ed0e537608cfdee7eed3ffa090d443 ipatests: tasks.py: fix ipa-epn invocation
* 41333b631d3ffa8d32d5bd5c2b69f2f51bb62c21 ipatests: test_epn: test_EPN_nbdays enhancements
